cdap-app-templates/cdap-etl/cdap-etl-realtime/src/main/java/co/cask/cdap/templates/etl/realtime/ETLRealtimeTemplate.java

```diff
@@ -0,0 +1,166 @@
+  private final Map<String, String> sinkClassMap;
+  private final Map<String, String> transformClassMap;
+  private RealtimeSource source;
+  private RealtimeSink sink;
+  private List<Transform> transforms;
+
+  public ETLRealtimeTemplate() throws Exception {
+    sourceClassMap = Maps.newHashMap();
+    sinkClassMap = Maps.newHashMap();
+    transformClassMap = Maps.newHashMap();
+    transforms = Lists.newArrayList();
+
+    initTable(Lists.<Class>newArrayList(IdentityTransform.class,
+                                        NoOpSink.class,
+                                        TestSource.class));
+  }
+
+  private void initTable(List<Class> classList) throws Exception {
+    for (Class klass : classList) {
+      DefaultStageConfigurer configurer = new DefaultStageConfigurer(klass);
+      if (RealtimeSource.class.isAssignableFrom(klass)) {
+        RealtimeSource source = (RealtimeSource) klass.newInstance();
+        source.configure(configurer);
+        sourceClassMap.put(configurer.createSpecification().getName(), configurer.createSpecification().getClassName());
+      } else if (RealtimeSink.class.isAssignableFrom(klass)) {
+        RealtimeSink sink = (RealtimeSink) klass.newInstance();
+        sink.configure(configurer);
+        sinkClassMap.put(configurer.createSpecification().getName(), configurer.createSpecification().getClassName());
+      } else {
+        Preconditions.checkArgument(Transform.class.isAssignableFrom(klass));
+        Transform transform = (Transform) klass.newInstance();
+        transform.configure(configurer);
+        transformClassMap.put(configurer.createSpecification().getName(),
+                              configurer.createSpecification().getClassName());
+      }
+    }
+  }
+
+  @Override
+  public void configureAdapter(String adapterName, ETLRealtimeConfig etlConfig, AdapterConfigurer configurer)
+    throws Exception {
+    int instances = etlConfig.getInstances();
+
+    ETLStage sourceConfig = etlConfig.getSource();
+    ETLStage sinkConfig = etlConfig.getSink();
+    List<ETLStage> transformConfigs = etlConfig.getTransforms();
+
+    // Instantiate Source, Transforms, Sink stages.
+    instantiateStages(sourceConfig, sinkConfig, transformConfigs);
+
+    // TODO: Validate Adapter
+
+    PipelineConfigurer pipelineConfigurer = new DefaultPipelineConfigurer(configurer);
+    configureSource(sourceConfig, configurer, pipelineConfigurer);
+    configureSink(sinkConfig, configurer, pipelineConfigurer);
+    configureTransforms(configurer);
+
+    configurer.addRuntimeArgument(Constants.ADAPTER_NAME, adapterName);
+    configurer.addRuntimeArgument(Constants.CONFIG_KEY, GSON.toJson(etlConfig));
+    configurer.setInstances(instances);
+  }
+
+  private void configureSource(ETLStage sourceConfig, AdapterConfigurer configurer,
+                               PipelineConfigurer pipelineConfigurer) throws Exception {
+    source.configurePipeline(sourceConfig, pipelineConfigurer);
```

Comment:
this stuff is very similar to batch. What if we have a common class 'Stage' or something like that, which has the configure() method that is shared by all sources, sinks, and transforms?  then you could re-use a lot of this code.